clash-lib/src/Clash/Driver/BrokenGhcs.hs

```diff
@@ -0,0 +1,156 @@
+{-|
+Copyright   :  (C) 2024, Martijn Bastiaan
+License     :  BSD2 (see the file LICENSE)
+Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+Utilities to detect and report GHC / operating system combinations that are
+known to be buggy.
+-}
+
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Clash.Driver.BrokenGhcs where
+
+import Data.Maybe (listToMaybe)
+import Data.Version (Version(Version, versionBranch))
+import GHC.Platform (OS(..))
+
+#if __GLASGOW_HASKELL__ > 810
+import System.Info (fullCompilerVersion)
+#endif
+
+import qualified Clash.Util.Interpolate as I
+import qualified System.Info
+
+#if __GLASGOW_HASKELL__ <= 810
+fullCompilerVersion :: Version
+fullCompilerVersion = System.Info.compilerVersion
+#endif
+
+-- | Current OS. Currently only recognizes Linux, Windows, and macOS.
+os :: OS
+os = case System.Info.os of
+  "darwin" -> OSDarwin
+  "linux" -> OSLinux
+  "mingw32" -> OSMinGW32
+  _ -> OSUnknown
+
+-- | What OS GHC is broken on (or all)
+data BrokenOn = All | SomeOs OS
+
+data GhcVersion
+  = Ghc Int Int Int
+  -- ^ @A.B.C@, for example @9.10.2@
+  deriving (Eq, Ord)
+
+data GhcRange = GhcRange
+  { from :: GhcVersion
+  -- ^ Start of range, inclusive
+  , to :: GhcVersion
+  -- ^ End of range, exclusive
+  }
+
+-- | Check if a 'GhcVersion' is within a 'GhcRange'
+ghcInRange :: GhcVersion -> GhcRange -> Bool
+ghcInRange ghc GhcRange{from, to} = from <= ghc && ghc < to
+
+-- | Construct a range of all GHC versions matching a major version
+ghcMajor :: Int -> Int -> GhcRange
+ghcMajor major minor = GhcRange{from=Ghc major minor 0, to=Ghc (major + 1) 0 0}
```

Comment:
This logic is wrong, and it is because of the mixed terminology. GHC has a 2-part major number. But here it's one part. So what happens is that
```haskell
ghcMajor 9 2
```
becomes
```haskell
GhcRange{from=Ghc 9 2 0, to=Ghc 10 0 0}
```

Reply:
🤦‍♂️ oh wow, I really fooled myself into that one

Reply:
Fixed?

Reply:
I think so. I don't think we need to worry about odd version numbers at all since that is a pretty weird setup, so no need to think about what we want to do with those anyway. And at the end of the day, the user could always use `-fclash-ignore-broken-ghcs`.

Reply:
The odd numbers are fine, given that the range's upper is exclusive.